src/compiler/checker.ts

```diff
@@ -16991,31 +16986,43 @@ namespace ts {
             return false;
         }
 
-        // Return true if type is of the given kind. A union type is of a given kind if all constituent types
-        // are of the given kind. An intersection type is of a given kind if at least one constituent type is
-        // of the given kind.
-        function isTypeOfKind(type: Type, kind: TypeFlags): boolean {
-            if (type.flags & kind) {
+        function isTypeOfKind(source: Type, kind: TypeFlags, excludeAny?: boolean) {

```

Comment:
`checkBinaryLikeExpression` has a complicated way of checking legal types for `+`:

1. It disallows `null + number` and `number + undefined`, even with `strictNullChecks` off. I like this check and want to keep it.
2. It gives a different types for `string + any ==> string` vs `T + any ==> any` and `T + unknown ==> unknown`.

It does this with the following (simplified) structure (using the abbreviation `tak` for `isTypeAssignableToKind`):

```ts
if (tak(left, NumberLike) && tak(right, NumberLike)) {
  return numberType;
}
else if (tak(left, StringLike) || tak(right, StringLike)) {
  return stringType;
}
else if (isTypeAny(left) || isTypeAny(right)) {
  return anyType;
}
else {
  returned undefined; // signal an error
}
```

Without `excludeAny` (now called `strict`), `isTypeAssignableTo(number, undefined)` is true, and `number + undefined/null` is now legal. Also, `isTypeAssignableTo(number, any)` is true, and `number + any ==> number` , and `any + any ==> number`.

I tried a few things to pull the `any` checks before the calls to `isTypeAssignableToKind`, but was not successful. If you see a way, let me know.